filter/conditional.go

```diff
@@ -0,0 +1,89 @@
+package filter
+
+import (
+	"github.com/silkeh/euler/filter/condition"
+	"github.com/silkeh/euler/util"
+	"sync"
+)
+
+// Conditional represents two parallel Filters and a condition.
+// If the condition is triggered on the conditionFilter:
+// Then n values from the dataFilter will be forwarded on the out channel.
+// Else n values will be discarded.
+type Conditional struct {
+	dataFilter      Filter
+	conditionFilter Filter
+	condition       condition.Condition
+	n               int
+}
+
+// NewConditional creates a new Conditional Filter
+func NewConditional(dataFilter Filter, conditionFilter Filter, condition condition.Condition, n int) Filter {
+	return &Conditional{
+		dataFilter:      dataFilter,
+		conditionFilter: conditionFilter,
+		condition:       condition,
+		n:               n,
+	}
+}
+
+// Run against an input and output channel until the input channel is closed.
+// The output channel is closed on return.
+func (c *Conditional) Run(in <-chan int, out chan<- int) {
+	var wg sync.WaitGroup
+	wg.Add(4)
+
+	defer func() {
+		wg.Wait()
```

Comment:
This should not be in the `defer`: a defer also runs on a `panic()`, in which case the `wg.Done()` may not be reached. This will then result in a hang.

Reply:
See: https://blog.golang.org/defer-panic-and-recover